src/jdk.jdwp.agent/share/native/libjdwp/classTrack.c

```diff
@@ -194,8 +212,9 @@ classTrack_initialize(JNIEnv *env)
 void
 classTrack_activate(JNIEnv *env)
 {
+    struct bag* new_bag = bagCreateBag(sizeof(char*), 1000);
```

Comment:
I don't think there can be any race during activation but this change is harmless.

Reply:
I agree. This code is called when the debug agent receives a CLASS_UNLOAD EventRequest (which the debug agent handles by requesting JVMTI GC_FINISH events). So there is nothing that prevents `classTrack_processUnloads()` from being called at the same time.

Reply:
In that case the `classTrack_processUnloads` is racing with the `classTrack_activate` call. I would not expect it to be possible to generate an event in relation to this before things have been "activated". But moving on ...

Reply:
Okay I see from later comments that in fact no such race exists, but we have to synchronize with cbTrackingObjectFree

Reply:
Bah! But then a comment after that says it does exist again.

There seem to be issues understanding exactly what the concurrent call sequences can be with this code.

Reply:
I think you might be confusing `classTrack_activate()` with `classTrack_reset()`. I believe the former has a race but the latter does not.

Reply:
...and even with the race, I still I think it is safe for a new `bagSize(deletedSignatures) == 0` check to be  added outside the lock. The only thing `classTrack_activate()` can do to `deletedSignatures` is set it when it is null. There is no risk of `deletedSignatures` being deleted and reallocated by another thread once you are in `classTrack_processUnloads()`. This is because it is called while holding the `handlerLock`, and as I explained elsewhere, by the time `classTrack_reset()` is called,  it's no longer possible to be in `classTrack_processUnloads()`.